main.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+)
+
+type project struct{}
+
+type Project interface {
+	ReadInput() (<-chan Command, []string)
+	Execute(cmds <-chan Command, input []string) []string
+	PrintResult(res []string)
+}
+
+var app Project
+
+type Command struct {
+	Name     string
+	Argument string
+}
+
+type Operation string
+
+const (
+	Cat  Operation = "cat"
+	Grep Operation = "grep"
+	Sort Operation = "sort"
+)
+
+func NewProject() Project {
+	return &project{}
+}
+
+func Run(p Project) {
+	cmds, out := p.ReadInput()
+	res := p.Execute(cmds, out)
+	p.PrintResult(res)
+}
+
+func (p *project) PrintResult(res []string) {
+	for _, line := range res {
+		fmt.Println(line)
+	}
+}
+
+func (p *project) ReadInput() (<-chan Command, []string) {
+	cmds, err := stdIn()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	var filePath string
+	out := make(chan Command, len(cmds))
+
+	go func(cmds []string) {
+		for _, cmd := range cmds {
+			sepCmd := strings.Fields(cmd)
+			if len(sepCmd) == 0 {
+				log.Fatal("input is not correct")
+			}
+			if len(sepCmd) < 2 && (filePath == "" || sepCmd[0] == string(Grep)) {
+				log.Fatal("input is not correct")
+			}
+
+			command := Command{}
+			command.Name = strings.ToLower(sepCmd[0])
+
+			if command.Name == string(Grep) {
+				command.Argument = sepCmd[1]
+			}
+
+			if filePath == "" {
+				filePath = sepCmd[len(sepCmd)-1]
+				wg.Done()
```

Comment:
better to use `defer wg.Done()`

Reply:
so using goroutine as well as waitgroup is not correct there, did i get it?